src/model/Publication.ts

```diff
@@ -247,4 +247,132 @@ export default class Publication {
+        positions.push(locator);
+        startPosition = startPosition + 1;
+      } else {
+        const href = this.getAbsoluteHref(link.href);
+        // Are we fetching every item in the readingOrder to generate positions?
+        // maybe in the future this can be done in the background?
+        const result = await fetch(href);
+        const length = (await result.blob()).size;
+        link.contentLength = length;
+        totalContentLength += length;
+        const positionLength = 1024;
+        const positionCount = Math.max(1, Math.ceil(length / positionLength));
+        Array.from(Array(positionCount).keys()).forEach((_, position) => {
+          const locator: Locator = {
+            href: link.href,
+            locations: {
+              progression: position / positionCount,
+              position: startPosition + (position + 1),
+            },
+            type: link.type,
+          };
+          positions.push(locator);
+        });
+        startPosition = startPosition + positionCount;
+      }
+
+      // If we are on the last item...
+      if (index + 1 === this.readingOrder.length) {
+        // for non fixed books, build up a dictionary of all the weights
+        // I am going to move this outside the map on readingOrder
+        // if (publication.metadata.rendition?.layout !== "fixed") {
+        //   publication.readingOrder.map(async (link) => {
+        //     link.contentWeight = (100 / totalContentLength) * link.contentLength;
+        //     weight[link.href] = link.contentWeight;
+        //   });
+        // }
+
+        // @aferditamuriqi what exactly does this do?
+        positions.map((locator, _index) => {
+          let resource = positions.filter(
+            (el: Locator) => el.href === decodeURI(locator.href)
+          );
+          let positionIndex = Math.ceil(
+            locator.locations.progression * (resource.length - 1)
+          );
+          locator.locations.totalProgression =
+            (locator.locations.position - 1) / positions.length;
+          locator.locations.remainingPositions = Math.abs(
+            positionIndex - (resource.length - 1)
+          );
+          locator.locations.totalRemainingPositions = Math.abs(
+            locator.locations.position - 1 - (positions.length - 1)
+          );
+        });
+        this.positions = positions;
+      }
+    });
+
+    // update the link.contentWeight to be a portion of the total and
+    // build up a map of link weights for non fixed layout publications
+    // @aferditamuriqi but what is the "weight" object ultimately used for?
+    const weight = {};
+    if (this.metadata.rendition?.layout !== "fixed") {
+      this.readingOrder.forEach((link) => {
+        // I don't totally know what this formula is saying haha, maybe we can
+        // add a comment
+        link.contentWeight = (100 / totalContentLength) * link.contentLength;
+        weight[link.href] = link.contentWeight;
+      });
+    }
```

Comment:
I pulled this out from inside the loop (you can see the old commented out version still in there). LMK if this looks okay? I couldn't really understand how the `weight` variable was used in the old `index.ts` code. Maybe you can help me out there? Would also be helpful to have a little comment on this block to know what it's meant to do.